src/tests/ftest/launch.py

```diff
@@ -239,11 +285,77 @@ def get_test_files(test_list):
     test_files = [{"py": test, "yaml": None} for test in test_list]
     for test_file in test_files:
         base, _ = os.path.splitext(test_file["py"])
-        test_file["yaml"] = "{}.yaml".format(base)
+        test_file["yaml"] = replace_yaml_file(
+            "{}.yaml".format(base), args, tmp_dir)
 
     return test_files
 
 
+def replace_yaml_file(yaml_file, args, tmp_dir):
+    """Replace the server/client yaml file placeholders.
+
+    Replace any server or client yaml file placeholder names with the host
+    names provided by the command line arguments in a copy of the original
+    test yaml file.  If no replacements are specified return the original
+    test yaml file.
+
+    Args:
+        yaml_file (str): test yaml file
+        args (argparse.Namespace): command line arguments for this program
+        tmp_dir (TemporaryDirectory): temporary directory object to use to
+            write modified yaml files
+
+    Returns:
+        str: the test yaml file
+
+    """
+    if args.test_servers:
+        # Determine which placeholder names need to be replaced in this yaml by
+        # getting the lists of hosts specified in the yaml file
+        unique_hosts = {"servers": set(), "clients": set()}
+        for key, placeholders in find_yaml_hosts(yaml_file).items():
+            if key in SERVER_KEYS:
+                unique_hosts["servers"].update(placeholders)
+            elif key in CLIENT_KEYS:
+                # If no specific clients are specified use a specified server
+                key = "clients" if args.test_clients else "servers"
+                unique_hosts[key].update(placeholders)
+
+        # Map the placeholder names to values provided by the user
+        mapping_pairings = [("servers", args.test_servers.split(","))]
+        if args.test_clients:
+            mapping_pairings.append(("clients", args.test_clients.split(",")))
+        mapping = {
+            tmp: node_list[index] if index < len(node_list) else None
+            for key, node_list in mapping_pairings
+            for index, tmp in enumerate(sorted(unique_hosts[key]))}
+
+        # Read in the contents of the yaml file to retain the !mux entries
+        print("Reading {}".format(yaml_file))
+        with open(yaml_file) as yaml_buffer:
+            file_str = yaml_buffer.read()
+
+        # Apply the placeholder replacements
+        for placeholder, host in mapping.items():
+            if host:
+                # Replace the host entries with their mapped values
+                file_str = re.sub(
+                    "- {}".format(placeholder), "- {}".format(host), file_str)
+            elif args.discard:
+                # Discard any host entries without a replacement value
+                file_str = re.sub(r"\s+- {}".format(placeholder), "", file_str)
```

Comment:
I know we've had this conversation before, but it seemed to [peter out](https://github.com/daos-stack/daos/pull/746#issuecomment-511881815) before any conclusion was reached.

So to resurrect it, should a test abort/fail here with a junit result indicating failure (i.e. rather than going on to run with invalid host names in the YAML file resulting in a failure result that needs some digging to figure out) if there were more placeholders in the original YAML file than there were hosts to replace them with and _-d_/_--discard_ was not given on the command line?

Reply:
I'll add an `else` clause.